File: polars_metrics/core/rolling.py
"""Rolling metrics for polars_metrics.

This module provides rolling (window-based) metrics that match QuantStats output.
"""

from __future__ import annotations

import math

import polars as pl

from polars_metrics.config import get_config
from polars_metrics.core.utils import (
    get_annualization_factor,
    to_float_series,
)
from polars_metrics.core.validation import validate_min_length, validate_returns


def rolling_volatility(
    returns: pl.Series,
    *,
    rolling_period: int = 126,
    periods_per_year: int | None = None,
    annualize: bool = True,
) -> pl.Series:
    """Calculate rolling volatility.

    Matches QuantStats rolling_volatility implementation.

    Parameters
    ----------
    returns : pl.Series
        Period returns.
    rolling_period : int, default 126
        Window size for rolling calculation (6 months for daily data).
    periods_per_year : int, optional
        Periods per year for annualization. If None, uses 365 to match QuantStats.
    annualize : bool, default True
        If True, annualize the volatility.

    Returns
    -------
    pl.Series
        Rolling annualized volatility series.

    Examples
    --------
    >>> import polars as pl
    >>> returns = pl.Series([0.01, -0.02, 0.015, -0.01, 0.02] * 50)
    >>> rolling_volatility(returns, rolling_period=10)
    shape: (250,)
    ...
    """
    validate_returns(returns, allow_empty=True)
    if returns.is_empty() or len(returns) < rolling_period:
        return pl.Series("rolling_volatility", [], dtype=pl.Float64)

    returns = to_float_series(returns)

    # QuantStats uses 365 by default for annualization in rolling functions
    ann_factor = get_annualization_factor(
        periods_per_year=periods_per_year or 365
    )

    # Calculate rolling std with ddof=1
    rolling_std = returns.rolling_std(window_size=rolling_period, ddof=1)

    if annualize:
        result = rolling_std * math.sqrt(ann_factor)
    else:
        result = rolling_std

    return result.alias("rolling_volatility")


def rolling_sharpe(
    returns: pl.Series,
    *,
    risk_free_rate: float = 0.0,
    rolling_period: int = 126,
    periods_per_year: int | None = None,
    annualize: bool = True,
) -> pl.Series:
    """Calculate rolling Sharpe ratio.

    Matches QuantStats rolling_sharpe implementation.

    Parameters
    ----------
    returns : pl.Series
        Period returns.
    risk_free_rate : float, default 0.0
        Annualized risk-free rate.
    rolling_period : int, default 126
        Window size for rolling calculation.
    periods_per_year : int, optional
        Periods per year for annualization. If None, uses 365 to match QuantStats.
    annualize : bool, default True
        If True, annualize the Sharpe ratio.

    Returns
    -------
    pl.Series
        Rolling Sharpe ratio series.

    Examples
    --------
    >>> import polars as pl
    >>> returns = pl.Series([0.01, -0.02, 0.015, -0.01, 0.02] * 50)
    >>> rolling_sharpe(returns, rolling_period=10)
    shape: (250,)
    ...
    """
    validate_returns(returns, allow_empty=True)
    if returns.is_empty() or len(returns) < rolling_period:
        return pl.Series("rolling_sharpe", [], dtype=pl.Float64)

    returns = to_float_series(returns)

    # QuantStats uses 365 by default for annualization in rolling functions
    ann_factor = get_annualization_factor(
        periods_per_year=periods_per_year or 365
    )

    # Convert annual risk-free to per-period
    rf_per_period = risk_free_rate / ann_factor

    # Excess returns
    excess_returns = returns - rf_per_period

    # Rolling mean of excess returns
    rolling_mean = excess_returns.rolling_mean(window_size=rolling_period)

    # Rolling std of returns (use original returns, not excess, like QuantStats)
    rolling_std = returns.rolling_std(window_size=rolling_period, ddof=1)

    # Sharpe = mean_excess / std_returns
    # Annualize: multiply by sqrt(ann_factor)
    if annualize:
        result = (rolling_mean / rolling_std) * math.sqrt(ann_factor)
    else:
        result = rolling_mean / rolling_std

    return result.alias("rolling_sharpe")


def rolling_sortino(
    returns: pl.Series,
    *,
    risk_free_rate: float = 0.0,
    rolling_period: int = 126,
    periods_per_year: int | None = None,
    annualize: bool = True,
) -> pl.Series:
    """Calculate rolling Sortino ratio.

    Matches QuantStats rolling_sortino implementation.

    Parameters
    ----------
    returns : pl.Series
        Period returns.
    risk_free_rate : float, default 0.0
        Annualized risk-free rate.
    rolling_period : int, default 126
        Window size for rolling calculation.
    periods_per_year : int, optional
        Periods per year for annualization. If None, uses 365 to match QuantStats.
    annualize : bool, default True
        If True, annualize the Sortino ratio.

    Returns
    -------
    pl.Series
        Rolling Sortino ratio series.

    Examples
    --------
    >>> import polars as pl
    >>> returns = pl.Series([0.01, -0.02, 0.015, -0.01, 0.02] * 50)
    >>> rolling_sortino(returns, rolling_period=10)
    shape: (250,)
    ...
    """
    validate_returns(returns, allow_empty=True)
    if returns.is_empty() or len(returns) < rolling_period:
        return pl.Series("rolling_sortino", [], dtype=pl.Float64)

    returns = to_float_series(returns)

    # QuantStats uses 365 by default for annualization
    ann_factor = get_annualization_factor(
        periods_per_year=periods_per_year or 365
    )

    # Convert annual risk-free to per-period
    rf_per_period = risk_free_rate / ann_factor

    # Adjust returns for risk-free rate
    adjusted = returns - rf_per_period

    # Rolling mean
    rolling_mean = adjusted.rolling_mean(window_size=rolling_period)

    # For downside deviation, we need custom rolling calculation
    # Use map_batches with a rolling window
    n = len(returns)
    downside_values = []

    for i in range(n):
        if i < rolling_period - 1:
            downside_values.append(None)
        else:
            window = adjusted[i - rolling_period + 1 : i + 1]
            # Get negative returns in window
            negative = window.filter(window < 0)
            if negative.is_empty():
                downside_values.append(0.0)
            else:
                # sqrt(sum(neg^2) / n)
                squared_sum = (negative ** 2).sum()
                downside = math.sqrt(squared_sum / rolling_period)
                downside_values.append(downside)

    downside_series = pl.Series("downside", downside_values)

    # Sortino = mean / downside * sqrt(ann_factor)
    if annualize:
        result = (rolling_mean / downside_series) * math.sqrt(ann_factor)
    else:
        result = rolling_mean / downside_series

    return result.alias("rolling_sortino")


def rolling_beta(
    returns: pl.Series,
    benchmark: pl.Series,
    *,
    rolling_period: int = 126,
) -> pl.Series:
    """Calculate rolling beta relative to benchmark.

    Parameters
    ----------
    returns : pl.Series
        Strategy period returns.
    benchmark : pl.Series
        Benchmark period returns.
    rolling_period : int, default 126
        Window size for rolling calculation.

    Returns
    -------
    pl.Series
        Rolling beta series.

    Examples
    --------
    >>> import polars as pl
    >>> returns = pl.Series([0.02, -0.01, 0.03, 0.01, -0.02] * 50)
    >>> benchmark = pl.Series([0.01, -0.02, 0.02, 0.01, -0.01] * 50)
    >>> rolling_beta(returns, benchmark, rolling_period=10)
    shape: (250,)
    ...
    """
    from polars_metrics.core.validation import validate_benchmark_match

    validate_returns(returns, allow_empty=True)
    validate_returns(benchmark, allow_empty=True)
    validate_benchmark_match(returns, benchmark)

    if returns.is_empty() or len(returns) < rolling_period:
        return pl.Series("rolling_beta", [], dtype=pl.Float64)

    returns = to_float_series(returns)
    benchmark = to_float_series(benchmark)

    n = len(returns)
    beta_values = []

    for i in range(n):
        if i < rolling_period - 1:
            beta_values.append(None)
        else:
            ret_window = returns[i - rolling_period + 1 : i + 1]
            bench_window = benchmark[i - rolling_period + 1 : i + 1]

            mean_ret = ret_window.mean()
            mean_bench = bench_window.mean()

            if mean_ret is None or mean_bench is None:
                beta_values.append(None)
                continue

            # Covariance
            cov = ((ret_window - mean_ret) * (bench_window - mean_bench)).mean()
            # Variance of benchmark
            var_bench = ((bench_window - mean_bench) ** 2).mean()

            if cov is None or var_bench is None or var_bench == 0:
                beta_values.append(None)
            else:
                beta_values.append(float(cov / var_bench))

    return pl.Series("rolling_beta", beta_values)

File: nanuquant-ts/src/core/rolling.ts
/**
 * Rolling metrics for nanuquant.
 *
 * This module provides rolling (window-based) metrics.
 *
 * Note: Rolling metrics default to 365 periods per year to match QuantStats.
 * To use trading days, pass `periodsPerYear: 252` explicitly.
 */

import { RollingGreeks } from '../types';
import { getAnnualizationFactor, mean, std, sum } from './utils';
import { validateReturns, validateBenchmarkMatch } from './validation';

/**
 * Calculate rolling standard deviation.
 *
 * @param arr - Input array.
 * @param windowSize - Size of rolling window.
 * @param ddof - Delta degrees of freedom. Default: 1.
 * @returns Array of rolling standard deviations (first windowSize-1 values are NaN).
 */
function rollingStd(arr: number[], windowSize: number, ddof: number = 1): number[] {
  const result: number[] = new Array(arr.length).fill(NaN);

  for (let i = windowSize - 1; i < arr.length; i++) {
    const window = arr.slice(i - windowSize + 1, i + 1);
    result[i] = std(window, ddof);
  }

  return result;
}

/**
 * Calculate rolling mean.
 *
 * @param arr - Input array.
 * @param windowSize - Size of rolling window.
 * @returns Array of rolling means (first windowSize-1 values are NaN).
 */
function rollingMean(arr: number[], windowSize: number): number[] {
  const result: number[] = new Array(arr.length).fill(NaN);

  for (let i = windowSize - 1; i < arr.length; i++) {
    const window = arr.slice(i - windowSize + 1, i + 1);
    result[i] = mean(window);
  }

  return result;
}

/**
 * Calculate rolling sum.
 *
 * @param arr - Input array.
 * @param windowSize - Size of rolling window.
 * @returns Array of rolling sums (first windowSize-1 values are NaN).
 */
function rollingSum(arr: number[], windowSize: number): number[] {
  const result: number[] = new Array(arr.length).fill(NaN);

  for (let i = windowSize - 1; i < arr.length; i++) {
    const window = arr.slice(i - windowSize + 1, i + 1);
    result[i] = sum(window);
  }

  return result;
}

/**
 * Calculate rolling volatility.
 *
 * @param returns - Period returns.
 * @param options - Optional parameters.
 * @param options.rollingPeriod - Window size. Default: 126 (6 months).
 * @param options.periodsPerYear - Periods per year for annualization. Default: 365.
 * @param options.annualize - Whether to annualize. Default: true.
 * @returns Array of rolling volatilities.
 *
 * @example
 * ```ts
 * const returns = [0.01, -0.02, 0.015, -0.01, 0.02, ...];
 * rollingVolatility(returns, { rollingPeriod: 10 });
 * ```
 */
export function rollingVolatility(
  returns: number[],
  options: {
    rollingPeriod?: number;
    periodsPerYear?: number;
    annualize?: boolean;
  } = {}
): number[] {
  const validated = validateReturns(returns, { allowEmpty: true });
  const {
    rollingPeriod = 126,
    periodsPerYear = 365,
    annualize = true,
  } = options;

  if (validated.length < rollingPeriod) {
    return [];
  }

  const annFactor = getAnnualizationFactor(periodsPerYear);
  const rollStd = rollingStd(validated, rollingPeriod, 1);

  if (annualize) {
    return rollStd.map((v) => (Number.isNaN(v) ? NaN : v * Math.sqrt(annFactor)));
  }

  return rollStd;
}

/**
 * Calculate rolling Sharpe ratio.
 *
 * @param returns - Period returns.
 * @param options - Optional parameters.
 * @param options.riskFreeRate - Annualized risk-free rate. Default: 0.
 * @param options.rollingPeriod - Window size. Default: 126.
 * @param options.periodsPerYear - Periods per year for annualization. Default: 365.
 * @param options.annualize - Whether to annualize. Default: true.
 * @returns Array of rolling Sharpe ratios.
 *
 * @example
 * ```ts
 * const returns = [0.01, -0.02, 0.015, -0.01, 0.02, ...];
 * rollingSharpe(returns, { rollingPeriod: 10 });
 * ```
 */
export function rollingSharpe(
  returns: number[],
  options: {
    riskFreeRate?: number;
    rollingPeriod?: number;
    periodsPerYear?: number;
    annualize?: boolean;
  } = {}
): number[] {
  const validated = validateReturns(returns, { allowEmpty: true });
  const {
    riskFreeRate = 0.0,
    rollingPeriod = 126,
    periodsPerYear = 365,
    annualize = true,
  } = options;

  if (validated.length < rollingPeriod) {
    return [];
  }

  const annFactor = getAnnualizationFactor(periodsPerYear);

  // Convert annual risk-free to per-period
  const rfPerPeriod = riskFreeRate / annFactor;

  // Excess returns
  const excessReturns = validated.map((r) => r - rfPerPeriod);

  // Rolling mean of excess returns
  const rollMean = rollingMean(excessReturns, rollingPeriod);

  // Rolling std of original returns (like QuantStats)
  const rollStd = rollingStd(validated, rollingPeriod, 1);

  // Sharpe = mean_excess / std_returns
  const result: number[] = [];
  for (let i = 0; i < validated.length; i++) {
    if (Number.isNaN(rollMean[i]) || Number.isNaN(rollStd[i]) || rollStd[i] === 0) {
      result.push(NaN);
    } else if (annualize) {
      result.push((rollMean[i] / rollStd[i]) * Math.sqrt(annFactor));
    } else {
      result.push(rollMean[i] / rollStd[i]);
    }
  }

  return result;
}

/**
 * Calculate rolling Sortino ratio.
 *
 * @param returns - Period returns.
 * @param options - Optional parameters.
 * @param options.riskFreeRate - Annualized risk-free rate. Default: 0.
 * @param options.rollingPeriod - Window size. Default: 126.
 * @param options.periodsPerYear - Periods per year for annualization. Default: 365.
 * @param options.annualize - Whether to annualize. Default: true.
 * @returns Array of rolling Sortino ratios.
 *
 * @example
 * ```ts
 * const returns = [0.01, -0.02, 0.015, -0.01, 0.02, ...];
 * rollingSortino(returns, { rollingPeriod: 10 });
 * ```
 */
export function rollingSortino(
  returns: number[],
  options: {
    riskFreeRate?: number;
    rollingPeriod?: number;
    periodsPerYear?: number;
    annualize?: boolean;
  } = {}
): number[] {
  const validated = validateReturns(returns, { allowEmpty: true });
  const {
    riskFreeRate = 0.0,
    rollingPeriod = 126,
    periodsPerYear = 365,
    annualize = true,
  } = options;

  if (validated.length < rollingPeriod) {
    return [];
  }

  const annFactor = getAnnualizationFactor(periodsPerYear);

  // Convert annual risk-free to per-period
  const rfPerPeriod = riskFreeRate / annFactor;

  // Adjusted returns
  const adjusted = validated.map((r) => r - rfPerPeriod);

  // Rolling mean of adjusted returns
  const rollMean = rollingMean(adjusted, rollingPeriod);

  // For downside deviation: square negative values, sum in window, divide by n, sqrt
  const negativeSquared = adjusted.map((r) => (r < 0 ? r * r : 0));
  const rollNegSqSum = rollingSum(negativeSquared, rollingPeriod);

  const result: number[] = [];
  for (let i = 0; i < validated.length; i++) {
    if (Number.isNaN(rollMean[i]) || Number.isNaN(rollNegSqSum[i])) {
      result.push(NaN);
    } else {
      const downside = Math.sqrt(rollNegSqSum[i] / rollingPeriod);
      if (downside === 0) {
        result.push(rollMean[i] > 0 ? Infinity : NaN);
      } else if (annualize) {
        result.push((rollMean[i] / downside) * Math.sqrt(annFactor));
      } else {
        result.push(rollMean[i] / downside);
      }
    }
  }

  return result;
}

/**
 * Calculate rolling beta relative to benchmark.
 *
 * @param returns - Strategy period returns.
 * @param benchmark - Benchmark period returns.
 * @param options - Optional parameters.
 * @param options.rollingPeriod - Window size. Default: 126.
 * @returns Array of rolling betas.
 *
 * @example
 * ```ts
 * const returns = [0.02, -0.01, 0.03, 0.01, -0.02, ...];
 * const benchmark = [0.01, -0.02, 0.02, 0.01, -0.01, ...];
 * rollingBeta(returns, benchmark, { rollingPeriod: 10 });
 * ```
 */
export function rollingBeta(
  returns: number[],
  benchmark: number[],
  options: { rollingPeriod?: number } = {}
): number[] {
  const validatedReturns = validateReturns(returns, { allowEmpty: true });
  const validatedBenchmark = validateReturns(benchmark, { allowEmpty: true });
  validateBenchmarkMatch(validatedReturns, validatedBenchmark);

  const { rollingPeriod = 126 } = options;

  if (validatedReturns.length < rollingPeriod) {
    return [];
  }

  // Product of returns and benchmark
  const product = validatedReturns.map((r, i) => r * validatedBenchmark[i]);

  // Benchmark squared
  const benchSq = validatedBenchmark.map((b) => b * b);

  // Rolling means
  const rollMeanRet = rollingMean(validatedReturns, rollingPeriod);
  const rollMeanBench = rollingMean(validatedBenchmark, rollingPeriod);
  const rollMeanProduct = rollingMean(product, rollingPeriod);
  const rollMeanBenchSq = rollingMean(benchSq, rollingPeriod);

  // Beta = Cov / Var
  // Cov = E[XY] - E[X]E[Y]
  // Var = E[Y^2] - E[Y]^2
  const result: number[] = [];
  for (let i = 0; i < validatedReturns.length; i++) {
    if (
      Number.isNaN(rollMeanRet[i]) ||
      Number.isNaN(rollMeanBench[i]) ||
      Number.isNaN(rollMeanProduct[i]) ||
      Number.isNaN(rollMeanBenchSq[i])
    ) {
      result.push(NaN);
    } else {
      const cov = rollMeanProduct[i] - rollMeanRet[i] * rollMeanBench[i];
      const varBench = rollMeanBenchSq[i] - rollMeanBench[i] * rollMeanBench[i];
      if (varBench === 0) {
        result.push(NaN);
      } else {
        result.push(cov / varBench);
      }
    }
  }

  return result;
}

/**
 * Calculate rolling alpha and beta relative to benchmark.
 *
 * @param returns - Strategy period returns.
 * @param benchmark - Benchmark period returns.
 * @param options - Optional parameters.
 * @param options.rollingPeriod - Window size. Default: 126.
 * @param options.periodsPerYear - Periods per year for alpha annualization. Default: 365.
 * @returns Object with alpha and beta arrays.
 *
 * @example
 * ```ts
 * const returns = [0.02, -0.01, 0.03, 0.01, -0.02, ...];
 * const benchmark = [0.01, -0.02, 0.02, 0.01, -0.01, ...];
 * const { alpha, beta } = rollingGreeks(returns, benchmark);
 * ```
 */
export function rollingGreeks(
  returns: number[],
  benchmark: number[],
  options: { rollingPeriod?: number; periodsPerYear?: number } = {}
): RollingGreeks {
  const validatedReturns = validateReturns(returns, { allowEmpty: true });
  const validatedBenchmark = validateReturns(benchmark, { allowEmpty: true });
  validateBenchmarkMatch(validatedReturns, validatedBenchmark);

  const { rollingPeriod = 126, periodsPerYear = 365 } = options;

  if (validatedReturns.length < rollingPeriod) {
    return { alpha: [], beta: [] };
  }

  const annFactor = getAnnualizationFactor(periodsPerYear);

  // Product of returns and benchmark
  const product = validatedReturns.map((r, i) => r * validatedBenchmark[i]);

  // Benchmark squared
  const benchSq = validatedBenchmark.map((b) => b * b);

  // Rolling means
  const rollMeanRet = rollingMean(validatedReturns, rollingPeriod);
  const rollMeanBench = rollingMean(validatedBenchmark, rollingPeriod);
  const rollMeanProduct = rollingMean(product, rollingPeriod);
  const rollMeanBenchSq = rollingMean(benchSq, rollingPeriod);

  const alphaResult: number[] = [];
  const betaResult: number[] = [];

  for (let i = 0; i < validatedReturns.length; i++) {
    if (
      Number.isNaN(rollMeanRet[i]) ||
      Number.isNaN(rollMeanBench[i]) ||
      Number.isNaN(rollMeanProduct[i]) ||
      Number.isNaN(rollMeanBenchSq[i])
    ) {
      alphaResult.push(NaN);
      betaResult.push(NaN);
    } else {
      const cov = rollMeanProduct[i] - rollMeanRet[i] * rollMeanBench[i];
      const varBench = rollMeanBenchSq[i] - rollMeanBench[i] * rollMeanBench[i];

      if (varBench === 0) {
        alphaResult.push(NaN);
        betaResult.push(NaN);
      } else {
        const beta = cov / varBench;
        const alphaPeriod = rollMeanRet[i] - beta * rollMeanBench[i];
        const alpha = alphaPeriod * annFactor;

        alphaResult.push(alpha);
        betaResult.push(beta);
      }
    }
  }

  return { alpha: alphaResult, beta: betaResult };
}
